cpp/include/cudf/types.hpp

```diff
@@ -295,7 +295,10 @@ class data_type {
  * @return true `lhs` is equal to `rhs`
  * @return false `lhs` is not equal to `rhs`
  */
-inline bool operator==(data_type const& lhs, data_type const& rhs) { return lhs.id() == rhs.id(); }
+inline bool operator==(data_type const& lhs, data_type const& rhs)
+{
+  return lhs.id() == rhs.id() and lhs.scale() == rhs.scale();
```

Comment:
This is the more canonical way to do multi-member comparison. 

```suggestion
  return std::tie(lhs.id(), lhs.scale()) == std::tie(rhs.id(), rhs.scale());
```

Reply:
Nice trick, I haven't seen this before.

Reply:
@jrhemstad 
Note, you have to make `operator==` a member function for the `std::tie` trick to work. Otherwise you get:
```
<source>: In function 'bool operator==(const data_type&, const data_type&)':
<source>:76:25: error: cannot bind non-const lvalue reference of type 'type_id&' to an rvalue of type 'type_id'
   76 |   return std::tie(lhs.id(), lhs.scale()) == std::tie(rhs.id(), rhs.scale());
```
As a side note, this was the error from GCC 10.2. Clang 11 gives you:
```
<source>:76:10: error: no matching function for call to 'tie'
  return std::tie(lhs.id(), lhs.scale()) == std::tie(rhs.id(), rhs.scale());
```
Note used to seeing better error messages from GCC.